File: in-progress-machine-learning-udemy/02-regression/simple-linear-regression.py
### Simple Linear Regression ###

## Data Preprocessing ##

# import libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.preprocessing import Imputer   # for missing data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder   # for categorical data
from sklearn.cross_validation import train_test_split   # for train test split
from sklearn.preprocessing import StandardScaler   # for features scaling

# import dataset
dataset = pd.read_csv('Salary_Data.csv')

# dependent variable - years experience
# independent variable - salary

X = dataset.iloc[:,0].values
y = dataset.iloc[:,1].values

# train test split
X_train, X_test, y_train, y_test = train_test_split(X, y,
                                                    test_size = float(1)/3,
                                                    random_state = 0)

# if one feature, need to do this
X_train = X_train.reshape(-1, 1)
X_test = X_test.reshape(-1, 1)

## Simple Linear Regression Model ##

# import libraries
from sklearn.linear_model import LinearRegression

# create LinearRegression object + fit
regressor = LinearRegression()
regressor.fit(X_train, y_train)

## Predicting the Test Set ##

y_pred = regressor.predict(X_test)

## Visualizing the Results ##

# training set results
plt.scatter(X_train, y_train, color = 'red')
# line is based on X_train and the model's predicted y values for X_train
plt.plot(X_train, regressor.predict(X_train), color = 'blue')
plt.title('Salary vs Experience (Training Set)')
plt.xlabel('Years of Experience')
plt.ylabel('Salary')
plt.show()   # ready to plot graph

# test set results
plt.scatter(X_test, y_test, color = 'red')
# same line
plt.plot(X_train, regressor.predict(X_train), color = 'blue')
plt.title('Salary vs Experience (Test Set)')
plt.xlabel('Years of Experience')
plt.ylabel('Salary')
plt.show()   # ready to plot graph

File: in-progress-machine-learning-udemy/02-regression/simple-linear-regression.R
### Simple Linear Regression ###

## Data Preprocessing ##

# import dataset
dataset = read.csv("Salary_Data.csv")

# independent variable - years of experience
# dependent variable - salary

# train test split
library(caTools)
set.seed(123)

split = sample.split(dataset$Salary, SplitRatio = 2/3)
train_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# feature scaling
# simple linear regression package takes care of feature scaling

## Simple Linear Regression ##

regressor = lm(formula = Salary ~ YearsExperience, data = train_set)
summary(regressor)

## Predicting Test Set Results ##

y_pred = predict(regressor, newdata = test_set)

## Visualize the Results ##

# install.packages('ggplot2')
library(ggplot2)

# training set results
ggplot() + geom_point(aes(x = train_set$YearsExperience, y = train_set$Salary),
                      color = 'red') +
  geom_line(aes(x = train_set$YearsExperience, y = predict(regressor, newdata = train_set)),
            color = 'blue') + 
  ggtitle('Salary vs Experience (Training Set)') +
  xlab('Years of Experience') +
  ylab('Salary')

# test set results
ggplot() + geom_point(aes(x = test_set$YearsExperience, y = test_set$Salary),
                      color = 'red') +
  geom_line(aes(x = train_set$YearsExperience, y = predict(regressor, newdata = train_set)),
            color = 'blue') + 
  ggtitle('Salary vs Experience (Test Set)') +
  xlab('Years of Experience') +
  ylab('Salary')
